### DISCUSSION TITLE
Development via Docker

### DISCUSSION THREAD

**Discussion by atomicbird** [srid/emanote#235] [Q&A]
Has anyone successfully set up this project for development in a Docker container? I wasn't sure I wanted to install nix on my desktop, so I'm trying to use an ubuntu-based Docker container combined with VS Code's remote development extension.

I started with a plain ubuntu image. Installing nix seems like it worked:

```
root@5c69b40feb67:/home/tph/emanote# nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.76-linuxkit, Ubuntu, 20.04.3 LTS (Focal Fossa)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.5.1`
 - channels(root): `"nixpkgs-22.05pre345869.46821ea01c8"`
 - nixpkgs: `/root/.nix-defexpr/channels/nixpkgs`
```

But then the language server gets killed:

```
root@5c69b40feb67:/home/tph/emanote# nix-shell --run haskell-language-server
haskell-language-server version: 1.5.1.0 (GHC: 8.10.7) (PATH: /nix/store/xxbjg2v8g4csvpjjd52yld6ly234asc1-haskell-language-server-1.5.1.0/bin/haskell-language-server)
 ghcide setup tester in /home/tph/emanote.
Report bugs at https://github.com/haskell/haskell-language-server/issues

Step 1/4: Finding files to test in /home/tph/emanote
Found 56 files

Step 2/4: Looking for hie.yaml files that control setup
Found 1 cradle
  (/home/tph/emanote/hie.yaml)

Step 3/4: Initializing the IDE

Step 4/4: Type checking the files
2022-01-14 23:02:57.088327877 [ThreadId 172] INFO hls:	Consulting the cradle for "src/Heist/Extra/Splices/Pandoc/TaskList.hs"
/tmp/nix-shell-1681-0/rc: line 3:  1739 Killed                  haskell-language-server
```

It doesn't give any clues about why it was killed, so I'm not sure what to do next. Is this a workable plan?

**Comment by srid** [comment]
Looks like HLS got killed after hitting some memory limit? Perhaps docker can be configured to increase that limit.

If you can get this working we can add a [`devcontainer.json` file](https://code.visualstudio.com/docs/remote/containers) to the repo. It would also enable people to hack on Emanote in the web browser using for example https://github.com/features/codespaces

**Reply by atomicbird** [reply]
That seems to be the case. Watching with `top` I can see the container run out of memory, then out of swap. I'm new to Docker so I'm not sure how to raise the limits beyond the options in the Docker desktop app, but I'll try to find out.

**Reply by srid** [reply]
`--memory=8G`? https://docs.docker.com/config/containers/resource_constraints/

**Reply by atomicbird** [reply]
It’s already at 16G. Since Docker desktop maxes out at 16 I don’t know if I can get past that.
